geth/chain.py

```diff
@@ -87,75 +100,32 @@ def is_ropsten_chain(data_dir: str) -> bool:
     return is_same_path(data_dir, get_ropsten_data_dir())
 
 
-# type ignored TODO rethink genesis file in a separate PR
-def write_genesis_file(  # type: ignore[no-untyped-def]
-    genesis_file_path,
-    overwrite=False,
-    nonce="0x0",
-    timestamp="0x0",
-    parentHash="0x0000000000000000000000000000000000000000000000000000000000000000",
-    extraData="0x0000000000000000000000000000000000000000000000000000000000000000",
-    gasLimit="0x47d5cc",
-    difficulty="0x0",
-    mixhash="0x0000000000000000000000000000000000000000000000000000000000000000",
-    coinbase="0x3333333333333333333333333333333333333333",
-    alloc=None,
-    config=None,
-):
+def write_genesis_file(
+    genesis_file_path: str,
+    overwrite: bool = False,
+    **genesis_data: Unpack[GenesisDataTypedDict],
+) -> None:
     if os.path.exists(genesis_file_path) and not overwrite:
         raise ValueError(
-            "Genesis file already present.  call with `overwrite=True` to overwrite this file"  # noqa: E501
+            "Genesis file already present. Call with "
+            "`overwrite=True` to overwrite this file"
         )
 
-    if alloc is None:
-        alloc = {}
-
-    if config is None:
-        config = {
-            "ethash": {},
-            "homesteadBlock": 0,
-            "daoForkBlock": 0,
-            "daoForkSupport": True,
-            "eip150Block": 0,
-            "eip155Block": 0,
-            "eip158Block": 0,
-            "byzantiumBlock": 0,
-            "constantinopleBlock": 0,
-            "petersburgBlock": 0,
-            "istanbulBlock": 0,
-            "berlinBlock": 0,
-            "londonBlock": 0,
-            "arrowGlacierBlock": 0,
-            "grayGlacierBlock": 0,
-            # merge
-            "terminalTotalDifficulty": 0,
-            "terminalTotalDifficultyPassed": True,
-            # post-merge, timestamp is used for network transitions
-            "shanghaiTime": 0,
-            "cancunTime": 0,
-        }
-
-    genesis_data = {
-        "nonce": nonce,
-        "timestamp": timestamp,
-        "parentHash": parentHash,
-        "extraData": extraData,
-        "gasLimit": gasLimit,
-        "difficulty": difficulty,
-        "mixhash": mixhash,
-        "coinbase": coinbase,
-        "alloc": alloc,
-        "config": config,
-    }
+    validate_genesis_data(genesis_data)
```

Comment:
I noticed this validation is done in `initialize_chain` and to separate concerncs, maybe this `write_genesis_file` method should only handle the file creation and writing of the data. The input data could be initialized and validated inside `intiailize_chain`.

Reply:
Yeah, I just added the validation wherever `genesis_data` was passed as an arg. But `write_genesis_file` is also a public function, so a user could just as easily call it without running through `initialize_chain` first. If the data is bad, it will still get caught and throw in the `genesis_data_filled = GenesisData(**genesis_data)` line, but having the call to `validate_genesis_data` would allow for consistent error messaging.